includes/oauth/class-mailchimp-api.php

```diff
@@ -143,37 +132,98 @@ private static function get_api_endpoint_from_key( $api_key ) {
-				],
-			]
-		);
-
+		$response = self::request( 'GET', '', [], $api_key );
 		if ( is_wp_error( $response ) ) {
 			return $response;
 		}
-		if ( 200 !== wp_remote_retrieve_response_code( $response ) ) {
-			$parsed_response = json_decode( $response['body'], true );
+		return \rest_ensure_response( [ 'username' => $response['username'] ] );
+	}
+
+	/**
+	 * Perform Mailchimp API Request
+	 *
+	 * @param string $method  HTTP method.
+	 * @param string $path    API path.
+	 * @param array  $data    Data to send.
+	 * @param string $api_key Optional API key to override the default one.
+	 *
+	 * @return array|WP_Error response body or error.
+	 */
+	public static function request( $method, $path = '', $data = [], $api_key = '' ) {
+		if ( empty( $api_key ) ) {
+			// 'newspack_mailchimp_api_key' is a new option introduced to manage MC API key accross Newspack plugins.
+			// Keeping the old option for backwards compatibility.
+			$api_key = \get_option( 'newspack_mailchimp_api_key', get_option( 'newspack_newsletters_mailchimp_api_key' ) );
+		}
+		$endpoint = self::get_api_endpoint_from_key( $api_key );
+		if ( ! $endpoint ) {
+			return new \WP_Error( 'wrong_parameter', __( 'Invalid Mailchimp API Key.', 'newspack' ) );
+		}
+		$url    = $endpoint . '/' . $path;
+		$config = [
+			'method'  => $method,
+			'headers' => [
+				'Accept'        => 'application/json',
+				'Content-Type'  => 'application/json',
+				'Authorization' => "Basic $api_key",
+			],
+		];
+		if ( ! empty( $data ) ) {
+			$config['body'] = \wp_json_encode( $data );
+		}
+		$response = \wp_safe_remote_request( $url, $config );
 
+		if ( \is_wp_error( $response ) ) {
+			return $response;
+		}
+		$parsed_response = json_decode( $response['body'], true );
+		if ( 200 !== \wp_remote_retrieve_response_code( $response ) ) {
 			return new \WP_Error(
 				'newspack_mailchimp_api',
 				array_key_exists( 'title', $parsed_response ) ? $parsed_response['title'] : __( 'Request failed.', 'newspack' )
 			);
 		}
+		return $parsed_response;
+	}
+
+	/**
+	 * Perform a GET request to Mailchimp's API
+	 *
+	 * @param string $path API path.
+	 *
+	 * @return array|WP_Error API response or error.
+	 */
+	public static function get( $path = '' ) {
+		return self::request( 'GET', $path );
+	}
 
-		$response_body = json_decode( $response['body'], true );
-		return \rest_ensure_response( [ 'username' => $response_body['username'] ] );
+	/**
+	 * Perform a PUT request to Mailchimp's API
+	 *
+	 * @param string $path API path.
+	 * @param array  $data Data to send.
+	 *
+	 * @return array|WP_Error API response or error.
+	 */
+	public static function put( $path = '', $data ) {
+		return self::request( 'PUT', $path, $data );
+	}
+
+	/**
+	 * Perform a POST request to Mailchimp's API
+	 *
+	 * @param string $path API path.
+	 * @param array  $data Data to send.
+	 *
+	 * @return array|WP_Error API response or error.
+	 */
+	public static function post( $path = '', $data ) {
```

Comment:
In PHP 8, required parameters must be listed first.

Reply:
In order to preserve the argument order like in the order methods, 7c715b361520426ca75a4ed6711d26c8768f4fd0 makes `$data` also optional